drivers/atwinc15x0: support dynamic scanning and connection to AP

fabian18 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

We currently only support static Wi-Fi credentials using `WIFI_SSID` and `WIFI_PASS`.
This PR implements a way of dynamic scanning and connecting to  Access Points using new netopts `NETOPT_SCAN`, `NETOPT_CONNECT`, and `NETOPT_DISCONNECT`.

- The `atwinc15x0` driver is adapted to support these options.
	- new pseudomodule `atwinc15x0_static_connect` can be activated to show the old behaviour
	- new pseudomodule `atwinc15x0_dynamic_connect` can be activated to support dynamic connection requests
	- new pseudomodule `atwinc15x0_dynamic_scan` can be activated to support dynamic scanning

You can enable whatever pseudomodules you want or even none.
	
- A `l2scan_list`data structure is added to deliver scan results in a list, sorted by their RSSI.

- A new shell command `iw` is added as an interface to interact with Wi-Fi interfaces
  
 
### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->

Test `l2scan_list`:
```
cd tests/l2scan_list
make
make term
```

```
RIOT native interrupts/signals initialized.
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2019.10-devel-19942-g71c2d-drivers_atwinc1510_scan_result)
...
OK (3 tests)
```

Test: `gnrc_networking`:

My wrapping Makefile:
<details>
<summary>Makefile_same54_xpro.mk</summary>

```Makefile
BOARD=same54-xpro

USEMODULE += shell_cmd_iw
USEMODULE += atwinc15x0
USEMODULE += atwinc15x0_static_connect
USEMODULE += atwinc15x0_dynamic_connect
USEMODULE += atwinc15x0_dynamic_scan
CFLAGS += -DDEBUG_ASSERT_VERBOSE=1

CFLAGS += -DATWINC15X0_PARAM_SSN_PIN="GPIO_PIN(1, 28)" \
          -DATWINC15X0_PARAM_RESET_PIN="GPIO_PIN(0, 6)" \
          -DATWINC15X0_PARAM_IRQ_PIN="GPIO_PIN(1, 7)" \
          -DATWINC15X0_PARAM_CHIP_EN_PIN="GPIO_PIN(0, 27)" \
          -DATWINC15X0_PARAM_WAKE_PIN="GPIO_PIN(0, 7)" \
          -DATWINC15X0_PARAM_SPI_CLK="MHZ(48)"

include Makefile
```

</details>

Try out `iw` command.
 
 
Test: `gnrc_border router`

My wrapping Makefile:
<details>
<summary>Makefile_same54_xpro.mk</summary>

```Makefile
BOARD ?= same54-xpro

CFLAGS += -DDEBUG_ASSERT_VERBOSE

UPLINK=wifi
CFLAGS += -DCONFIG_GNRC_DHCPV6_CLIENT_6LBR_UPSTREAM=5

USEMODULE += shell_cmd_iw
USEMODULE+=atwinc15x0_dynamic_connect
USEMODULE+=atwinc15x0_dynamic_scan
USEMODULE+=atwinc15x0
CFLAGS += -DATWINC15X0_PARAM_SSN_PIN="GPIO_PIN(1, 28)" \
          -DATWINC15X0_PARAM_RESET_PIN="GPIO_PIN(0, 6)" \
          -DATWINC15X0_PARAM_IRQ_PIN="GPIO_PIN(1, 7)" \
          -DATWINC15X0_PARAM_CHIP_EN_PIN="GPIO_PIN(0, 27)" \
          -DATWINC15X0_PARAM_WAKE_PIN="GPIO_PIN(0, 7)" \
          -DATWINC15X0_PARAM_SPI_CLK="MHZ(48)"

include Makefile
```

</details>

```
2023-03-14 12:28:09,044 # main(): This is RIOT! (Version: 2019.10-devel-19942-g71c2d-drivers_atwinc1510_scan_result)
2023-03-14 12:28:09,045 # RIOT border router example application
2023-03-14 12:28:09,045 # All up, running the shell now
iw 5 connect ... -k ...
2023-03-14 12:28:23,940 # iw 5 connect ... -k ...
2023-03-14 12:28:24,056 # [atwinc15x0] WiFi connected
2023-03-14 12:28:24,061 # connected to ...
2023-03-14 12:28:24,062 # iw: ok
> ifconfig
2023-03-14 12:28:33,190 # ifconfig
2023-03-14 12:28:33,221 # Iface  5  HWaddr: F8:F0:05:AD:B2:94  Channel: 0  RSSI: -52  Link: up 
2023-03-14 12:28:33,223 #            State: IDLE 
2023-03-14 12:28:33,227 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2023-03-14 12:28:33,230 #           Source address length: 6
2023-03-14 12:28:33,233 #           Link type: wireless
2023-03-14 12:28:33,240 #           inet6 addr: 2a0c:d242:4482:3a00:faf0:5ff:fead:b294  scope: global  VAL
2023-03-14 12:28:33,243 #           inet6 group: ff02::2
2023-03-14 12:28:33,245 #           inet6 group: ff02::1
2023-03-14 12:28:33,249 #           inet6 group: ff02::1:ffad:b294
2023-03-14 12:28:33,250 #           
2023-03-14 12:28:33,253 # Iface  6  HWaddr: FC:C2:3D:2F:8E:90 
2023-03-14 12:28:33,258 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2023-03-14 12:28:33,261 #           Source address length: 6
2023-03-14 12:28:33,263 #           Link type: wired
2023-03-14 12:28:33,269 #           inet6 addr: fe80::fec2:3dff:fe2f:8e90  scope: link  VAL
2023-03-14 12:28:33,272 #           inet6 group: ff02::2
2023-03-14 12:28:33,274 #           inet6 group: ff02::1
2023-03-14 12:28:33,278 #           inet6 group: ff02::1:ff2f:8e90
2023-03-14 12:28:33,279 #
```

Somehow the uplink does not configure a link-local address.
I will investigate on this problem.

### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

gschorcht commented on sys/include/net/wifi.h:
```diff
@@ -0,0 +1,207 @@
+ */
+#define WIFI_2_4_CH_NUMOF               (2 + (WIFI_2_4_CH_MAX - WIFI_2_4_CH_MIN))
+
+/**
+ * @brief   Type to express a Wi-Fi security mode
+ */
+typedef int8_t wifi_security_mode_t;
+
+/**
+ * @brief   The AP is unprotected
+ */
+#define WIFI_SECURITY_MODE_NONE         0
+
+/**
+ * @brief   The AP is protected with WEP
+ */
+#define WIFI_SECURITY_MODE_WEP_PSK      1
+
+/**
+ * @brief   The AP is protected with WPA
+ */
+#define WIFI_SECURITY_MODE_WPA_PSK      2
+
+/**
+ * @brief   The AP requires user authentication with username and password
+ */
+#define WIFI_SECURITY_MODE_EAP_USER     3
```
Please use enumaration type:
```suggestion
/**
 * @brief   Type to express a Wi-Fi security mode
 */
enum {
    WIFI_SECURITY_MODE_NONE,          /**<  AP is unprotected */
    ...
} wifi_security_mode_t;
```

gschorcht replied:
Not yet address :wink:

fabian18 replied:
[42f76bd](https://github.com/RIOT-OS/RIOT/pull/19387/commits/42f76bde2e1a9182477296ff0e1318486c5ad6ec) ?

fabian18 replied:
respective renaming was done in other fixups because they fixup different commits

gschorcht replied:
Ah ok, I didn't see this update when I started to review again, sorry.

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -144,6 +308,137 @@ typedef union {
 
 static bool _rssi_info_ready = false;
 
+static void _atwinc15x0_handle_M2M_WIFI_RESP_SCAN_DONE(tstrM2mScanDone* scan_done)
```
Please use lower case function names here an below. Function names should be a bit shorter not to hit the 100 characters limit, for examples:
```suggestion
static void _atwinc15x0_handle_scan_done(tstrM2mScanDone* scan_done)
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -144,6 +308,137 @@ typedef union {
+    }
+}
+
+static void _atwinc15x0_handle_M2M_WIFI_RESP_SCAN_RESULT(const tstrM2mWifiscanResult* scan_result)
```
```suggestion
static void _atwinc15x0_handle_scan_result(const tstrM2mWifiscanResult* scan_result)
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -144,6 +308,137 @@ typedef union {
+#if IS_USED(MODULE_ATWINC15X0_DYNAMIC_SCAN)
+    wifi_security_mode_t sec;
+    if ((sec = _atwinc15x0_get_sec_mode(scan_result->u8AuthType)) != -1) {
+        wifi_scan_result_t result = WIFI_SCAN_RESULT_INITIALIZER(scan_result->u8ch, scan_result->s8rssi, sec);
```
```suggestion
        wifi_scan_result_t result = WIFI_SCAN_RESULT_INITIALIZER(scan_result->u8ch,
                                                                 scan_result->s8rssi, sec);
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -144,6 +308,137 @@ typedef union {
+    }
+}
+
+static void _atwinc15x0_handle_M2M_WIFI_RESP_CON_STATE_CHANGED(const tstrM2mWifiStateChanged *state_changed)
```
```suggestion
static void _atwinc15x0_handle_con_state_changed(const tstrM2mWifiStateChanged *state_changed)
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -144,6 +308,137 @@ typedef union {
+    }
+}
+
+static void _atwinc15x0_handle_M2M_WIFI_RESP_CONN_INFO(const tstrM2MConnInfo *conn_info)
```
```suggestion
static void _atwinc15x0_handle_conn_info(const tstrM2MConnInfo *conn_info)
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -144,6 +308,137 @@ typedef union {
+#endif
+}
+
+static void _atwinc15x0_handle_M2M_WIFI_RESP_CURRENT_RSSI(int8_t rssi)
```
```suggestion
static void _atwinc15x0_handle_current_rssi(int8_t rssi)
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -157,85 +452,19 @@ static void _atwinc15x0_wifi_cb(uint8_t type, void *msg)
-                  event->scan_done.u8NumofCh);
-            /* read the first scan result record */
-            m2m_wifi_req_scan_result(0);
+            _atwinc15x0_handle_M2M_WIFI_RESP_SCAN_DONE(&event->scan_done);
```
```suggestion
            _atwinc15x0_handle_resp_scan_done(&event->scan_done);
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -157,85 +452,19 @@ static void _atwinc15x0_wifi_cb(uint8_t type, void *msg)
-                /* read the next scan result record */
-                m2m_wifi_req_scan_result(event->scan_result.u8index + 1);
-            }
+            _atwinc15x0_handle_M2M_WIFI_RESP_SCAN_RESULT(&event->scan_result);
```
```suggestion
            _atwinc15x0_handle_resp_scan_result(&event->scan_result);
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -157,85 +452,19 @@ static void _atwinc15x0_wifi_cb(uint8_t type, void *msg)
-                default:
-                    break;
-            }
+            _atwinc15x0_handle_M2M_WIFI_RESP_CON_STATE_CHANGED(&event->state_changed);
```
```suggestion
            _atwinc15x0_handle_resp_con_state_changed(&event->state_changed);
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -157,85 +452,19 @@ static void _atwinc15x0_wifi_cb(uint8_t type, void *msg)
-            atwinc15x0->rssi = event->conn_info.s8RSSI;
-            memcpy(atwinc15x0->ap,
-                   event->conn_info.au8MACAddress, ETHERNET_ADDR_LEN);
+            _atwinc15x0_handle_M2M_WIFI_RESP_CONN_INFO(&event->conn_info);
```
```suggestion
            _atwinc15x0_handle_resp_conn_info(&event->conn_info);
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -157,85 +452,19 @@ static void _atwinc15x0_wifi_cb(uint8_t type, void *msg)
-            break;
-
-        default:
+            _atwinc15x0_handle_M2M_WIFI_RESP_CURRENT_RSSI(event->rssi);
```
```suggestion
            _atwinc15x0_handle_resp_current_rssi(event->rssi);
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -144,6 +308,137 @@ typedef union {
+    /* set the RSSI and BSSID of the current AP */
+    atwinc15x0->rssi = conn_info->s8RSSI;
+    memcpy(atwinc15x0->ap, conn_info->au8MACAddress, ETHERNET_ADDR_LEN);
+#if IS_USED(MODULE_ATWINC15X0_DYNAMIC_CONNECT)
+    strcpy(atwinc15x0->ssid, conn_info->acSSID);
+#endif
```
```suggestion
    if (IS_USED(MODULE_ATWINC15X0_DYNAMIC_CONNECT)) {
        strcpy(atwinc15x0->ssid, conn_info->acSSID);
    }
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -144,6 +308,137 @@ typedef union {
+    else {
+        /* no results */
+        _atwinc15x0_set_scanning(atwinc15x0, false);
+#if IS_USED(MODULE_ATWINC15X0_DYNAMIC_SCAN)
+        wifi_scan_list_t result = WIFI_SCAN_LIST_INITIALIZER(NULL);
+        _wifi_scan_result_callback(&result);
+#endif
```
Is it possible to use the following here?
```suggestion
        if (IS_USED(MODULE_ATWINC15X0_DYNAMIC_SCAN)) {
            wifi_scan_list_t result = WIFI_SCAN_LIST_INITIALIZER(NULL);
            _wifi_scan_result_callback(&result);
        }
```

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -144,6 +308,137 @@ typedef union {
+            atwinc15x0->channel = scan_result->u8ch;
+        }
+    }
+#if IS_USED(MODULE_ATWINC15X0_DYNAMIC_SCAN)
+    wifi_security_mode_t sec;
+    if ((sec = _atwinc15x0_get_sec_mode(scan_result->u8AuthType)) != -1) {
+        wifi_scan_result_t result = WIFI_SCAN_RESULT_INITIALIZER(scan_result->u8ch, scan_result->s8rssi, sec);
+        memcpy(result.bssid, scan_result->au8BSSID, sizeof(result.bssid));
+        strncpy(result.ssid, (char *)scan_result->au8SSID, sizeof(result.ssid) - 1);
+        _wifi_scan_list_insert(&result);
+    }
+#endif
```
Possible here?
```suggestion
    if (IS_USED(MODULE_ATWINC15X0_DYNAMIC_SCAN)) {
        wifi_security_mode_t sec;
        if ((sec = _atwinc15x0_get_sec_mode(scan_result->u8AuthType)) != -1) {
            wifi_scan_result_t result = WIFI_SCAN_RESULT_INITIALIZER(scan_result->u8ch,
                                                                     scan_result->s8rssi, sec);
            memcpy(result.bssid, scan_result->au8BSSID, sizeof(result.bssid));
            strncpy(result.ssid, (char *)scan_result->au8SSID, sizeof(result.ssid) - 1);
            _wifi_scan_list_insert(&result);
        }
    }
```

gschorcht reviewed (changes requested):
Some first comments.

gschorcht commented:
Thanks for contribution this feature.

Some additional remarks:
- You have a lot of preprocessor conditionals `#if IS_USED(...)` in your code. I stopped commenting them. Please check which could be used as C conditionals `if (IS_USED(...= {` instead.
- WPA2 Enterprise mode should never be called `WIFI_SECURITY_MODE_EAP_USER` but  `WIFI_SECURITY_MODE_WPA2_ENTERPRISE`.
- `EAP user and password are used for inner authentification in WPA2 Enterprise mode. For the outer authentification you could usually define a identity. Is there something in the driver for using that?

gschorcht commented:
BTW, we should have this feature also in our prime WiFi platform ESP32x :wink:

gschorcht commented on drivers/atwinc15x0/Makefile.dep:
```diff
@@ -1,6 +1,14 @@
 USEMODULE += ztimer_msec
 USEMODULE += netdev_legacy_api
+
+ifneq (,$(filter atwinc15x0_dynamic_scan,$(USEMODULE)))
```
When I tried to test this PR the first time, I realized that `examples/gnrc_networking` doesn't work out of the box anymore. To get it working, module `atwinc15x0_static_connect` has to be enabled explicitly. Therefore, I suggest to enable `atwinc15x0_static_connect` by default if no other module is enabled:
```suggestion
ifeq (,$(filter atwinc15x0_dynamic_%,$(USEMODULE)))
  USEMODULE += atwinc15x0_static_connect
endif

ifneq (,$(filter atwinc15x0_dynamic_scan,$(USEMODULE)))
```

gschorcht replied:
What about this change to enable `atwinc15x0_static_connect` if none of the `atwinc15x0_dynamic_*` modules is enabled?

gschorcht replied:
> It´s [574191e](https://github.com/RIOT-OS/RIOT/commit/574191e583343d222d84b8ce2961af5fcdbfb1f7)
> 
> Are fixups getting messy? Do you want me to squash?

Ok

gschorcht replied:
> It´s [574191e](https://github.com/RIOT-OS/RIOT/commit/574191e583343d222d84b8ce2961af5fcdbfb1f7)

Sorry again for asking again for this change. Now I know why I thought that this issue wasn't fix. Usually, GitHub labels files for which a new version is uploaded as `Outdated`. However, this file isn't labeled as outdated and if you click on the file name, it still openes the initial version. That's why I didn't realized that you changed it already.

gschorcht commented on sys/shell/cmds/iw.c:
```diff
@@ -0,0 +1,394 @@
+
+static void _list_ap(void)
+{
+    puts("                            SSID |         SEC |  RSSI | CHANNEL");
```
Why are SSID and SEC column right aligned?

fabian18 replied:
There was no particular reason for that it is just what e.g. `%32s` does. Do you want me to left-align the columns with `%-32s`?

gschorcht replied:
It looks a bit strange, should be right aligned but names/labels a.s.o should be left aligned. Yes, please change it.

benpicco commented:
Your eMail address looks a bit mangled - you want to run

    git config --global user.email

on that machine.

benpicco commented on drivers/atwinc15x0/include/atwinc15x0_internal.h:
```diff
@@ -34,6 +37,136 @@ extern "C" {
+ * @brief   Flag to indicate that the transceiver is scanning at the moment
+ *          and is thus not able to start a new scan concurrently
+ */
+#define ATWINC15X0_FLAG_SCANNING            (((atwinc15x0_flags_t)1) << 0)
+
+/**
+ * @brief   Flag to indicate that the transceiver is trying to connect to an AP
+ *          at the moment and is thus not able to process another connect
+ *          request concurrently
+ */
+#define ATWINC1510_FLAG_CONNECTING          (((atwinc15x0_flags_t)1) << 1)
+
+/**
+ * @brief   Flag to indicate that the transceiver is connected to an AP
+ */
+#define ATWINC15X0_FLAG_CONNECTED           (((atwinc15x0_flags_t)1) << 2)
```
Can those be set at the same time?
From a glance it feels like distinct states like

```C
typedef enum {
    ATWINC1510_STATE_DISCONNECTED,
    ATWINC1510_STATE_DISCONNECTED_SCANNING,
    ATWINC1510_STATE_CONNECTING,
    ATWINC1510_STATE_CONNECTED,
    ATWINC1510_STATE_CONNECTED_SCANNING,
} atwinc15x0_state_t;
``` 

would be easier to handler.

fabian18 replied:
Since
```C
    if (IS_USED(MODULE_ATWINC15X0_STATIC_CONNECT)) {
        /* try to connect and return */
        res = _atwinc15x0_static_connect();
    }
    if (IS_USED(MODULE_ATWINC15X0_DYNAMIC_SCAN)) {
        wifi_scan_request_t request = WIFI_SCAN_REQUEST_INITIALIZER(NETOPT_SCAN_REQ_ALL_CH,
                                                                    _wifi_scan_cb, 0);
        _atwinc15x0_scan(&request);
    }
```
did not work when both modules were enabled, `ATWINC1510_STATE_CONNECTING_SCANNING` would not be a valid state, so I think your state analysis is right. Maybe it requires one more `is_connected` check to distinguish between  `ATWINC1510_STATE_DISCONNECTED_SCANNING` and `ATWINC1510_STATE_CONNECTED_SCANNING`, but I can try to reformulate it with states.

fabian18 replied:
I am trying to turn `netopt_state_t state` into `atwinc15x0_state_t state` in `atwinc15x0_t`. It is a bit more effort than I thought, to think about what operation can be performed in which state.

fabian18 replied:
I must still test a bit with the state handling

benpicco commented on sys/include/net/netopt.h:
```diff
@@ -878,6 +905,133 @@ typedef enum {
+ */
+typedef struct netopt_connect_request {
+    netopt_on_connect_result_t conn_cb;         /**< On connect callback */
+    netopt_on_disconnect_result_t disconn_cb;   /**< On disconnect callback */
```
I see some redundancy with `NETDEV_EVENT_LINK_UP`/`NETDEV_EVENT_LINK_DOWN` with those - could #17902 be a generic alternative here?

fabian18 replied:
An alternative maybe, but not sure.
An improvement for interested threads who did not request the action definately!
Giving a callback is not enforced, and also I would not exclude the case to send a message in the callback.
If #17902 gets merged before, I would also try with this, but I´m not sure if both ways are mutually exclusive.

fabian18 commented:
> git config --global user.email

Thank you for the hint. I applied the command with my new email.

fabian18 commented:
> EAP user and password are used for inner authentification in WPA2 Enterprise mode. For the outer authentification you could usually define a identity. Is there something in the driver for using that?

In `m2m_types.h`
```C
typedef enum {
	M2M_WIFI_SEC_INVALID = 0,
	/*!< Invalid security type.
	*/
	M2M_WIFI_SEC_OPEN,
	/*!< Wi-Fi network is not secured.
	*/
	M2M_WIFI_SEC_WPA_PSK,
	/*!< Wi-Fi network is secured with WPA/WPA2 personal(PSK).
	*/
	M2M_WIFI_SEC_WEP,
	/*!< Security type WEP (40 or 104) OPEN OR SHARED.
	*/
	M2M_WIFI_SEC_802_1X
	/*!< Wi-Fi network is secured with WPA/WPA2 Enterprise.IEEE802.1x user-name/password authentication.
	 */
}tenuM2mSecType;
```

```C
#define M2M_1X_USR_NAME_MAX								21
/*!< The maximum size of the user name including the NULL termination.
	It is used for RADIUS authentication in case of connecting the device to
	an AP secured with WPA-Enterprise.
*/


#define M2M_1X_PWD_MAX									41
/*!< The maximum size of the password including the NULL termination.
	It is used for RADIUS authentication in case of connecting the device to
	an AP secured with WPA-Enterprise.
*/
```

```C
/*!
@struct	\
	tstr1xAuthCredentials

@brief
	Credentials for the user to authenticate with the AAA server (WPA-Enterprise Mode IEEE802.1x).
*/
typedef struct{
	uint8	au8UserName[M2M_1X_USR_NAME_MAX];
	/*!< User Name. It must be Null terminated string.
	*/
	uint8	au8Passwd[M2M_1X_PWD_MAX];
	/*!< Password corresponding to the user name. It must be Null terminated string.
	*/
}tstr1xAuthCredentials;

```

```C
/*!
@union	\
	tuniM2MWifiAuth

@brief
	Wi-Fi Security Parameters for all supported security modes.
*/
typedef union{
	uint8				au8PSK[M2M_MAX_PSK_LEN];
	/*!< Pre-Shared Key in case of WPA-Personal security.
	*/
	tstr1xAuthCredentials	strCred1x;
	/*!< Credentials for RADIUS server authentication in case of WPA-Enterprise security.
	*/
	tstrM2mWifiWepParams	strWepInfo;
	/*!< WEP key parameters in case of WEP security.
	*/
}tuniM2MWifiAuth;
```

I am not so much familiar with WPA2 Enterprise but I have read that it can also deal with certificates and in your original PR you said you were having problems to connect to an Enterprise AP. I put the latest FW 19.7.7 on my module so maybe it works. I have not tried to connect to an Enterprise AP so far, but I think we have one in the office.

fabian18 commented:
@gschorcht I think I addressed your comments.
Some stats:

```
# static connect only
   text    data     bss     dec     hex filename
  99876     252   31936  132064   203e0 /home/fabian.huessler@ml-pa.loc/RIOT/examples/gnrc_networking/bin/same54-xpro/gnrc_networking.elf
```
![cosy_static_connect_only](https://user-images.githubusercontent.com/15147337/225915178-a62b77b7-07a8-49cd-bb2a-992304228334.png)

```
# static connect and dynamic connect
   text    data     bss     dec     hex filename
 100388     252   32000  132640   20620 /home/fabian.huessler@ml-pa.loc/RIOT/examples/gnrc_networking/bin/same54-xpro/gnrc_networking.elf
 ```
![cosy_static_connect_and_dynamic_connect](https://user-images.githubusercontent.com/15147337/225915531-4ba1b656-db51-4c81-a2ce-99db83615786.png)

```
 # static connect and dynamic connect and dynamic scan
    text    data     bss     dec     hex filename
 100648     252   32032  132932   20744 /home/fabian.huessler@ml-pa.loc/RIOT/examples/gnrc_networking/bin/same54-xpro/gnrc_networking.elf
```
![cosy_static_connect_dynamic_connect_dynamic_scan](https://user-images.githubusercontent.com/15147337/225915641-a5c86b9b-8511-4b12-acfc-d350a1cba32d.png)


@benpicco I will try to change the flags to states and have a look at #17902

gschorcht commented on sys/include/net/wifi.h:
```diff
@@ -0,0 +1,207 @@
+
+/**
+ * @brief   2.4 GHz channel 7
+ */
+#define WIFI_2_4_CH_7                   7
+
+/**
+ * @brief   2.4 GHz channel 8
+ */
+#define WIFI_2_4_CH_8                   8
+
+/**
+ * @brief   2.4 GHz channel 9
+ */
+#define WIFI_2_4_CH_9                   9
+
+/**
+ * @brief   2.4 GHz channel 10
+ */
+#define WIFI_2_4_CH_10                  10
+
+/**
+ * @brief   2.4 GHz channel 11
+ */
+#define WIFI_2_4_CH_11                  11
+
+/**
+ * @brief   2.4 GHz channel 12
+ */
+#define WIFI_2_4_CH_12                  12
+
+/**
+ * @brief   2.4 GHz channel 13
+ */
+#define WIFI_2_4_CH_13                  13
+
+/**
+ * @brief   2.4 GHz channel 14
+ */
+#define WIFI_2_4_CH_14                  14
```
Please use whenever possible enumaration for constant definitions.
```suggestion
/**
 * @brief   2.4 GHz Channels
 */
enum {
    WIFI_2_4_CH_1 = 1       /**< 2.4 GHz channel 1 */
    WIFI_2_4_CH_2 = 2       /**< 2.4 GHz channel 2 */
    WIFI_2_4_CH_3 = 3       /**< 2.4 GHz channel 3 */
    WIFI_2_4_CH_4 = 4       /**< 2.4 GHz channel 4 */
    WIFI_2_4_CH_5 = 5       /**< 2.4 GHz channel 5 */
    WIFI_2_4_CH_6 = 6       /**< 2.4 GHz channel 6 */
    WIFI_2_4_CH_7 = 7       /**< 2.4 GHz channel 7 */
    WIFI_2_4_CH_8 = 8       /**< 2.4 GHz channel 8 */
    WIFI_2_4_CH_9 = 9       /**< 2.4 GHz channel 9 */
    WIFI_2_4_CH_10 = 10     /**< 2.4 GHz channel 10 */
    WIFI_2_4_CH_11 = 11     /**< 2.4 GHz channel 11 */
    WIFI_2_4_CH_12 = 12     /**< 2.4 GHz channel 12 */
    WIFI_2_4_CH_13 = 13     /**< 2.4 GHz channel 13 */
    WIFI_2_4_CH_14 = 14     /**< 2.4 GHz channel 14 */
};
```

gschorcht commented on sys/shell/cmds/iw.c:
```diff
@@ -0,0 +1,394 @@
+    puts("                            SSID |         SEC |  RSSI | CHANNEL");
+    puts("---------------------------------+-------------+-------+--------");
+    for (unsigned i = 0; i < _aps.numof; i++) {
+        printf("%32s |%12s |%6"PRId16" |%7d \n",
```
```suggestion
        printf("%-32s |%-12s |%6"PRId16" |%7d \n",
```

gschorcht commented:
> In `m2m_types.h`
> 
> ```c
> typedef enum {
> 	M2M_WIFI_SEC_INVALID = 0,
> 	/*!< Invalid security type.
> 	*/
> 	M2M_WIFI_SEC_OPEN,
> 	/*!< Wi-Fi network is not secured.
> 	*/
> 	M2M_WIFI_SEC_WPA_PSK,
> 	/*!< Wi-Fi network is secured with WPA/WPA2 personal(PSK).
> 	*/
> 	M2M_WIFI_SEC_WEP,
> 	/*!< Security type WEP (40 or 104) OPEN OR SHARED.
> 	*/
> 	M2M_WIFI_SEC_802_1X
> 	/*!< Wi-Fi network is secured with WPA/WPA2 Enterprise.IEEE802.1x user-name/password authentication.
> 	 */
> }tenuM2mSecType;
> ```

Independent on what the ATWINC1500 M2M API defines, we should use the terminology as defined by the WiFi Alliance.

The IEEE 802.11 doesn't know the term WPA/WPA2 Personal Mode but only the encryption protocols TKIP, GCCP and GCMP that use a pre-shared key (or passphrase) . What's called WPA2 Enterprise Mode was defined in IEEE 802.11i as RSN which is just a IEEE 802.1x-based authentication over any kind of TLS tunnel. That's why the security type is called `M2M_WIFI_SEC_802_1X` here.

EAP (IEEE 802.1x) was specified at the time when it was very usual that a user terminal (called supplicant in EAP terminology) used dial-in line via PPP to connect to a network. The supplicant had to authenicate via EAP at the dial-in gateway (called authenticator) which used a RADIUS or Kerberos authentication server. EAP is also used for authentication at Ethernet ports of switches that support the port-based access control according to IEEE 802.1x.

WPA2 Enterprise Mode just reuses the EAP over any kind of TLS tunnel and is therefore called IEEE 802.1x authentication by the IEEE. I would therefore suggest to use the WiFi-Alliance terminology (WPA2 Personal and WPA Enterprise) in our API.

- [a.] EAP-TLS = EAP Transport Layer Security
- [b.] EAP-TTLS = EAP Tunneled Transport Layer Security
- [c.] PEAP = Protected EAP

Variant a. is the most complicated one because it requires a client certificate which is quite difficult to handle in embedded systems. Variant b. and c. use 2 phases of authentications:
- Phase 1
  Firtst, the client (supplicant) uses the EAP to connect the authentication server via an unencrypted connection. Based on the so-called anonymous identity, the authentication server specifies which kind of TLS tunnel and which authentication method has to be used in phase 2. The client then establishes a TLS connection with the authentication server.
  This process is also called **outer** EAP authentication.
- Phase 2 
  The client uses again the EAP but now via the encrypted TLS connection to authenticate against the authentication server with the authentication method specified by the authentication server.
  This process is then called **inner** EAP authentication.

Although the `atwinc5x0_netdev` API doesn't allow to specify the anonymous identity in phase 1, we should take it into account in the our interface to be able to use it also for other WiFi netdevs than the `atwinc5x0_netdev`, e.g. `esp_wif`.

gschorcht commented on drivers/atwinc15x0/atwinc15x0_netdev.c:
```diff
@@ -482,11 +727,47 @@ static int _atwinc15x0_set(netdev_t *netdev, netopt_t opt, const void *val,
                 return max_len;
             }
+        case NETOPT_SCAN:
+            if (!IS_USED(MODULE_ATWINC15X0_DYNAMIC_SCAN)) { goto forward_to_netdev; }
```
Since all cases return directly, the `goto` statement can be replaces easily by a `break` and calling 
```c
return netdev_eth_set(netdev, opt, val, max_len);
```
after the the switch/case construct.

gschorcht commented on drivers/include/atwinc15x0.h:
```diff
@@ -54,16 +63,30 @@ typedef struct {
+    ATWINC15X0_STATE_DISCONNECTED_CONNECTING,   /**< Connecting state where disconnected before */
+    ATWINC15X0_STATE_CONNECTED,                 /**< Connected state */
+    ATWINC15X0_STATE_CONNECTED_SCANNING,        /**< Scanning state when connected */
+    ATWINC15X0_STATE_CONNECTED_CONNECTING,      /**< Connecting state where disconnect event is not yet received */
```
Exceeding the maximum line length of 100 characters
```suggestion
    ATWINC15X0_STATE_CONNECTED_CONNECTING,      /**< Connecting state where disconnect event is not yet received */
```

gschorcht commented on sys/include/net/netopt.h:
```diff
@@ -803,6 +806,30 @@ typedef enum {
+     * The result is a list of reachable access points @ref l2scan_list_t.
+     * Notification happens by a callback @ref netopt_on_scan_result_t.
+     */
+    NETOPT_SCAN,
```
Are these options (`SCAN`, `CONNECT` and `DISCONNECT`) unique for WiFi? If yes, I wonder whether we should call them `NEOPT_WIFI_*`.

fabian18 replied:
I also see them for 802.15.4 when a node wants to connect to a PAN coordinator

fabian18 replied:
I am not sure if there is active scanning in 802.15.4 but receiving of beacons exist at least

fabian18 commented:
> Somehow the uplink does not configure a link-local address.
I will investigate on this problem.

It happens sometimes, when `set(NETOPT_L2_GROUP)` --> `m2m_wifi_enable_mac_mcast()` fails with `M2M_ERR_MEM_ALLOC`.

fabian18 commented:
or `M2M_ERR_BUS_FAIL`

gschorcht commented:
> > Somehow the uplink does not configure a link-local address.
> > I will investigate on this problem.
> 
> It happens sometimes, when `set(NETOPT_L2_GROUP)` --> `m2m_wifi_enable_mac_mcast()` fails with `M2M_ERR_MEM_ALLOC`.

Anything new on this problem?

fabian18 commented:
Yes, 81154fe351852a94a3ee7d58fdd2f1191413a64e helps here. With the current version it works fine to switch between connection, disconnection, scanning, sleeping states. However in another branch I want to build a network manager around it which sends itself messages in the callback and has the task to dispatch the events to subscribers. I can use the netif bus for this I hope. So far while developing I found things I want to restructure here a bit and I will push it when I think that the global picture is good.

gschorcht commented on tests/l2scan_list/Makefile:
```diff
@@ -0,0 +1,6 @@
+include ../Makefile.tests_common
```
The test has to be moved to `tests/net` directory.

gschorcht replied:
Otherwise, the PR looks good to me. Are there any issues open? My test with a `nucleo-f411re` and the Adafruit WINC1500 Arduino shield worked as expected including GUA and LLUA.

gschorcht replied:
Once you have moved the test app to `tests/net`, please rebase and solve the conflict.

fabian18 replied:
Thanks for looking at it again.

> Are there any issues open?

Not as fat as I can tell. We also started to use the updated driver with still the static variant in production.
Nevertheless I found some tiny things worth to be fixed up.

One thing that I noted after a rebase to master is a pin conflict in my configuration on the `same54-xpro`.
The Ethernet IRQ pin (PD12) uses the same external interrupt controller (EXTINT[7]) as the EXT1 pin header interrupt pin (PB7) on the same54.
As a result no interrupt events had been generated. So if anyone uses this board, the EXT2 pin header should be used.
```
ATWINC15X0_PARAM_EXT2 = "{                                \
                        .spi         = SPI_DEV(1),        \
                        .spi_clk     = SPI_CLK_10MHZ,     \
                        .ssn_pin     = GPIO_PIN(2, 6),    \
                        .reset_pin   = GPIO_PIN(1, 1),    \
                        .irq_pin     = GPIO_PIN(3, 0),    \
                        .chip_en_pin = GPIO_PIN(1, 2),    \
                        .wake_pin    = GPIO_PIN(1, 6),    \
                      }" 
```

gschorcht commented on drivers/atwinc15x0/Makefile.dep:
```diff
@@ -1,6 +1,18 @@
 USEMODULE += netdev_legacy_api
+
+ifeq (,$(filter atwinc15x0_dynamic_%,$(USEMODULE)))
+# use static connect by default when no dynamic module is loaded
```
```suggestion
  # use static connect by default when no dynamic module is loaded
```

gschorcht reviewed (approved):
Looks good an works as expected.

gschorcht commented:
Please squash. Use a shorter commit message for commit f932a3b2cd7ff3d7d3bf5f24000d23d7e62aabe9, it exhausts the maximum of 72 characters.

fabian18 commented:
Awesome, thanks!

fabian18 commented:
@gschorcht are you fine with the force pushed changes which showed to be necessary from the Murdock run?

gschorcht commented:
> @gschorcht are you fine with the force pushed changes which showed to be necessary from the Murdock run?

Yes, I'm fine with squashing changes directly to make Murdock happy.

gschorcht commented:
Maybe @benpicco can merge it together with other PRs that are already approved. I don't know whether setting the label `CI: read for merge train` is enough. BTW, the label seems to have a typo, I guess it should be `CI: ready for merge train`.

benpicco commented:
The merge process is still very manual, re-triggering #19253 gives some time to write

bors merge

beneath all PRs that should be merged. The label is just to ease filtering for PRs to include in the merge train.
